### PR TITLE
Update DEVELOPER_GUIDE.md instructions for JDK-11

### PR DESCRIPTION
**Pull request by reta**
### Description
Update DEVELOPER_GUIDE.md instructions for JDK-11, fixes the annoying deprecation failure:

```
Executing Gradle on JVM versions 16 and lower has been deprecated. This will fail with an error in Gradle 9.0. Use JVM 17 or greater to execute Gradle. Projects can continue to use older JVM versions via toolchains. Consult the upgrading guide for further information: https://docs.gradle.org/8.10.2/userguide/upgrading_version_8.html#minimum_daemon_jvm_version


> Configure project :
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 8.10.2
  OS Info               : Linux 6.11.0-9-generic (amd64)
  JDK Version           : 11 (Ubuntu JDK)
  JAVA_HOME             : /usr/lib/jvm/java-11-openjdk-amd64
  Random Testing Seed   : AE535EC00EACC589
  In FIPS 140 mode      : false
=======================================

FAILURE: Build failed with an exception.

* What went wrong:
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0

```

### Related Issues
N/A (popped up in Slack from community member)

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -92,7 +92,13 @@ sourceCompatibility = JavaVersion.VERSION_11
...
-Download Java 11 from [here](https://adoptium.net/releases.html?variant=openjdk11).
+Download Java 11 from [here](https://adoptium.net/releases.html?variant=openjdk11). 
+
+Since Gradle has deprecated the usage of the JDKs below JDK-16, we recommend to use [JDK 17](#jdk-17) for development. If you still want to stick to JDK-11, please add `-Dorg.gradle.warning.mode=none` when invoking any Gradle build task from command line, for example:
+
+```
+./gradlew check -Dorg.gradle.warning.mode=none
+```
 
 #### JDK 14
```
Comment (andrross) [inline DEVELOPER_GUIDE.md:103]: Is this section still accurate? I haven't had a JDK-14 installed in a long long time and I don't think we should encourage folks to go out and download a version that has been EOL for so long.

**Review by andrross — COMMENTED** [review]
Lots of things out of date here. Some file references are now stale (`buildSrc/version.properties`). The comment about JDK 17 being the bundled JDK is wrong. Could we simplify this to say something to the effect of: "We recommend installing the latest LTS JDK and setting it as `JAVA_HOME`. To run the full suite of bwc tests you must also install and configure `JAVA11_HOME`."

**Comment by reta** [comment]
> Lots of things out of date here. Some file references are now stale (`buildSrc/version.properties`). The comment about JDK 17 being the bundled JDK is wrong.

It is, was focusing on immediate "need" but you are 100% right, going over the whole document

File: DEVELOPER_GUIDE.md (hunk 1 above)
Comment (reta) [reply DEVELOPER_GUIDE.md]: Dropped it altogether, thanks @andrross !

File: DEVELOPER_GUIDE.md [hunk 2]
```diff
@@ -92,19 +91,21 @@ sourceCompatibility = JavaVersion.VERSION_11
...
 
-#### JDK 14
+Since 8.10 release, Gradle has deprecated the usage of the JDKs below JDK-16. As such, for development purposes we recommend to install additionally at least [JDK 17](https://adoptium.net/temurin/releases/?version=17) or [JDK 21](https://adoptium.net/temurin/releases/?version=21). If you still want to stick to JDK-11 only, please add `-Dorg.gradle.warning.mode=none` when invoking any Gradle build task from command line, for example:
 
-To run the full suite of tests, download and install [JDK 14](https://jdk.java.net/archive/) and set `JAVA11_HOME`, and `JAVA14_HOME`. They are required by the [backwards compatibility test](./TESTING.md#testing-backwards-compatibility).
+```
+./gradlew check -Dorg.gradle.warning.mode=none
+```
 
-#### JDK 17
```
Comment (owaiskazi19) [inline DEVELOPER_GUIDE.md:101]: Should we add JDK17 back as @andrross called out we need 17 for bwc?

Comment (reta) [reply DEVELOPER_GUIDE.md:101]: On the second thought, I think JDK-17 should probably stay there, thanks @owaiskazi19 !

**Review by owaiskazi19 — APPROVED** [review]

File: DEVELOPER_GUIDE.md (hunk 2 above)
Comment (andrross) [reply DEVELOPER_GUIDE.md:101]: I'm wondering if we could simplify the three "JDK" headings down to just this:

> \#### JDK
> 
> OpenSearch requires a minimum of JDK-11 in order to build. However, it is recommended to use a more recent LTS version of the JDK and set it to the `JAVA_HOME` environment variable. In addition, certain backward compatibility tests check out and compile the previous major version of OpenSearch, and therefore require installing JDK-11 and JDK-17 and setting the `JAVA11_HOME` and `JAVA17_HOME` environment variables.
> 
> By default, the test tasks use bundled JDK at runtime, which is configured in version catalog [gradle/libs.versions.toml](gradle/libs.versions.toml).

The existing text has a lot of detail that I think is maybe not necessary?

Comment (reta) [reply DEVELOPER_GUIDE.md:101]: > I'm wondering if we could simplify the three "JDK" headings down to just this:
> 
> > #### JDK
> > OpenSearch requires a minimum of JDK-11 in order to build. However, it is recommended to use a more recent LTS version of the JDK and set it to the `JAVA_HOME` environment variable. In addition, certain backward compatibility tests check out and compile the previous major version of OpenSearch, and therefore require installing JDK-11 and JDK-17 and setting the `JAVA11_HOME` and `JAVA17_HOME` environment variables.
> > By default, the test tasks use bundled JDK at runtime, which is configured in version catalog [gradle/libs.versions.toml](gradle/libs.versions.toml).
> 
> The existing text has a lot of detail that I think is maybe not necessary?

There are a few things here that do not play well (hence this pull request):
 - due to Gradle, to build with any JDK below 16, one needs to add `-Dorg.gradle.warning.mode=none`
 - due to Gradle, build with any JDK above 23 will fail  (like 24-ea for example)

So I think we should be very explicit what we support or not, so I could introduce the build matrix instead. Also, the large chunk of this work will be throwaway very soon (after https://github.com/opensearch-project/OpenSearch/pull/16366). 

If you don't mind, I would prefer to keep this guide exact (with respect to JDK versions) and get back to it once we level up to JDK-21.

Comment (andrross) [reply DEVELOPER_GUIDE.md:101]: Sure, that's fair, but what I don't like about the text in this PR is towards the beginning of this section it says:

> This means you must have a JDK 11 installed with the environment variable `JAVA_HOME` referencing the path to Java home for your JDK 11 installation, e.g. `JAVA_HOME=/usr/lib/jvm/jdk-11`.

Only if you keep reading does it recommend to use Java 17 or 21. And nowhere does it mention the JAVA11_HOME and JAVA17_HOME environment variables which are required for bwc tests. I also don't love the snippets of gradle configuration which is destined to be out of date when things change anyway.

Comment (reta) [reply DEVELOPER_GUIDE.md:101]: Valid points, reworked these sections, thanks @andrross !

File: DEVELOPER_GUIDE.md [hunk 3]
```diff
@@ -76,35 +74,24 @@ Fork [opensearch-project/OpenSearch](https://github.com/opensearch-project/OpenS
 
 ### Install Prerequisites
 
-#### JDK 11
+#### JDK
 
-OpenSearch builds using Java 11 at a minimum, using the Adoptium distribution. This means you must have a JDK 11 installed with the environment variable `JAVA_HOME` referencing the path to Java home for your JDK 11 installation, e.g. `JAVA_HOME=/usr/lib/jvm/jdk-11`. This is configured in [buildSrc/build.gradle](buildSrc/build.gradle) and [distribution/tools/java-version-checker/build.gradle](distribution/tools/java-version-checker/build.gradle).
+OpenSearch builds using Java 11 at a minimum, using the [Temurin/Adoptium](https://adoptium.net/temurin/releases/) distribution. This means you must have a JDK 11 installed with the environment variable `JAVA_HOME` referencing the path to Java home for your JDK 11 installation, e.g. `JAVA_HOME=/usr/lib/jvm/jdk-11`. This is configured in [buildSrc/build.gradle](buildSrc/build.gradle) and [distribution/tools/java-version-checker/build.gradle](distribution/tools/java-version-checker/build.gradle).
```
Comment (andrross) [inline DEVELOPER_GUIDE.md:79]: What do you think about the following?

```suggestion
OpenSearch recommends building with the [Temurin/Adoptium](https://adoptium.net/temurin/releases/) distribution. JDK 11 is the minimum supported, and JDK-23 is the newest supported. You must have a supported JDK installed with the environment variable `JAVA_HOME` referencing the path to Java home for your JDK installation, e.g. `JAVA_HOME=/usr/lib/jvm/jdk-21`. 
```

Comment (reta) [reply DEVELOPER_GUIDE.md]: Like it, thanks @andrross !

**Comment by reta** [comment]
@andrross LGTY? thanks!

**Review by andrross — APPROVED** [review]

**Pull request merged**